src/gui/dialogs/migrate_version_selection.cpp

```diff
@@ -97,19 +99,52 @@ void migrate_version_selection::post_show(window& window)
 		// given self-compilation and linux distros being able to do whatever they want plus command line options to
 		// alter locations make sure the directories/files are actually different before doing anything with them
 		if(migrate_addons_dir != filesystem::get_addons_dir()) {
+			std::vector<std::string> old_addons;
+			std::vector<std::string> current_addons;
 			std::vector<std::string> migrate_addons;
-			filesystem::get_files_in_dir(migrate_addons_dir, nullptr, &migrate_addons);
+
+			filesystem::get_files_in_dir(migrate_addons_dir, nullptr, &old_addons);
+			filesystem::get_files_in_dir(filesystem::get_addons_dir(), nullptr, &current_addons);
+
+			std::set_difference(old_addons.begin(), old_addons.end(), current_addons.begin(), current_addons.end(), std::back_inserter(migrate_addons));
+
 			if(migrate_addons.size() > 0) {
 				ad_hoc_addon_fetch_session(migrate_addons);
 			}
 		}
 
 		if(migrate_prefs_file != filesystem::get_prefs_file() && filesystem::file_exists(migrate_prefs_file)) {
-			filesystem::copy_file(migrate_prefs_file, filesystem::get_prefs_file());
+			// if this is the first time, just copy the file over
+			// else need to merge the preferences file
+			if(first_time_) {
+				filesystem::copy_file(migrate_prefs_file, filesystem::get_prefs_file());
+			} else {
+				config current_cfg;
+				read(current_cfg, filesystem::get_prefs_file());
+				config old_cfg;
+				read(old_cfg, migrate_prefs_file);
+
+				// when both files have the same attribute, use the one from whichever was most recently modified
+				bool current_prefs_are_older = filesystem::file_modified_time(filesystem::get_prefs_file()) < filesystem::file_modified_time(migrate_prefs_file);
+				for(const config::attribute& val : old_cfg.attribute_range()) {
+					if((current_cfg.has_attribute(val.first) && !current_prefs_are_older) || !current_cfg.has_attribute(val.first)) {
+						preferences::set(val.first, val.second);
+					}
+				}
+
+				for(const auto& val : old_cfg.all_children_range()) {
+					// only move tags that exist in the older version's file but not in the current version's file
+					// can't blindly overwrite since that could give undesirable results
+					// ie: for achievements and campaign completions
+					if(!current_cfg.has_child(val.key)) {
+						preferences::set_child(val.key, val.cfg);
+					}
+				}
+			}
 		}
 
-		if(migrate_credentials_file != filesystem::get_credentials_file()
-			&& filesystem::file_exists(migrate_credentials_file)) {
+		// don't touch the credentials file on migrator re-run
+		if(migrate_credentials_file != filesystem::get_credentials_file() && filesystem::file_exists(migrate_credentials_file) && first_time_) {
```

Comment:
Could check if the current credentials file exists instead?

Reply:
That wouldn't accomplish anything given the credentials file is created as soon as you close wesnoth for the first time. And if someone went to the trouble to manually delete their credentials file, then they can surely copy the old one over themselves too.

Reply:
But that sounds like it would work exactly like with the first_time bool? It seems to me like all the cases that are handled with that bool can be handled "naturally" as well. So we could get rid of it.

Reply:
Updated.